[JENKINS-50447] New design and slim down of login page

scherler opened the pull request:
The change will provide a new designed login page that is using no external deps to prevent any possible malicious javascript. 

## new design

![screenshot_2018-04-12_14-41-41](https://user-images.githubusercontent.com/596701/38677814-a3f8688c-3e5f-11e8-8392-5b5af5cd48c4.png)


See [JENKINS-50447](https://issues.jenkins-ci.org/browse/JENKINS-50447).

### Proposed changelog entries

* Entry 1: JENKINS-50447, new designed login page that is using no external deps to prevent any possible malicious javascript.

### Proposed changelog (@oleg-nenashev)

* Major RFE, JENKINS-50447: Introduce new design of the Login page
* RFE, JENKINS-50447: Introduce new SimplePageDecorator extension point, 
which allows decorating the Login Page

Compatibility notes: The new Login page UI will no longer accept common `PageDecorator` extension points, so plugins like Simple Theme Plugin will no longer be able to apply changes to the page.
`SimplePageDecorator` extension point should be used instead 

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/code-reviewers @reviewbybees @cloudbees/team-astro

scherler committed:
[JENKINS-50447] New design and slim down of login page

The change will provide a new designed login page that is using no external deps to prevent any possible malvious javascript.
We still allow the user to use the legacy login page, which has not changed the behaviour.

[JENKINS-50447] Implement error case and use final svg jenkins logo

[JENKINS-50447] remove mock up logo

[JENKINS-50447] Fix feet back from jlong

[JENKINS-50447_squash] Use theme css which only contains values to be overriden. Drop ie js since with the theme css it is not needed anymore

[JENKINS-50447_squash] Fix expression to evaluate when to show an error and when to simply redirect

eslint - formating changes and fix offences

[JENKINS-50447_squash] Use system property to controll whether or not to use the legacy login

[JENKINS-50447_squash] prevent that the jsessionId appears in the url which should fix most tests

[JENKINS-50447_squash] Fix cli test by adding pageDecorators magic from layout.jelly

[JENKINS-50447_squash] remove pageDecorators because they assume existence of js that we do not have

[JENKINS-50447_squash] make login page extensible with pagedecorators

[JENKINS-50447_squash] drop legacy support

scherler committed:
[JENKINS-50447] New design and slim down of login page

The change will provide a new designed login page that is using no external deps to prevent any possible malvious javascript.
We still allow the user to use the legacy login page, which has not changed the behaviour.

[JENKINS-50447] Implement error case and use final svg jenkins logo

[JENKINS-50447] remove mock up logo

[JENKINS-50447] Fix feet back from jlong

[JENKINS-50447_squash] Use theme css which only contains values to be overriden. Drop ie js since with the theme css it is not needed anymore

kzantow commented:
I think this is going to be _awesome_! Could you also provide a screenshot(s)?

scherler commented:
@kzantow cheers, I attached now screenshots from latest FF. I can add some other browser if needed but all should look like that (well IE11 is a bit off, but yeah)

oleg-nenashev commented:
* Do you retain the footer with Jenkins information. It may be critical for a user since it displays App Name and Version
* Maybe user sign-up page also needs to be updated.

The change may also hit UX on corporate instances which apply custom styles using Simple Theme Plugin. Applying corporate look&feel to a new page will require patches. It is not a big deal since everything is in CSS tho.

Also CC @recena taking the nature of changes

jtnord commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,103 @@ THE SOFTWARE.
+    <j:otherwise>
+      <st:setHeader name="Expires" value="0" />
+      <st:setHeader name="Cache-Control" value="no-cache,no-store,must-revalidate" />
+      <st:setHeader name="X-Hudson-Theme" value="default" />
```
What is the purpose of this header?

scherler replied:
I have to admit I am not sure. It had been there before so I did not removed it. However I may want to change the value to `fresh`

daniel-beck replied:
https://github.com/jenkinsci-cert/jenkins/commit/7bee9394389de54ed402ddf0ccf1df14dbbeda8a is the only reference I could find. No related issues from around that time turned up. We could probably just remove this header in a separate PR and see what happens.

jtnord commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,103 @@ THE SOFTWARE.
+      <j:set var="inputClass" value="${error ? 'danger' : 'normal'}"/>
+      <html>
+        <head data-rooturl="${rootURL}" data-resurl="${resURL}" resURL="${resURL}">
+          <title>${h.appendIfNotNull('Login', ' [Jenkins]', 'Jenkins')}</title>
```
Should "Jenkins" not use the name from the war?

scherler replied:
As well this is coming from the original, where this had been coming via `j:layout`

jtnord commented on war/src/main/webapp/css/login.theme.css:
```diff
@@ -0,0 +1,53 @@
+    color: #4A4A4A;
+}
+.logo {
+    background-image: url('../images/jenkins.svg');
```
does IE support this as a background?

scherler replied:
![screenshot_2018-04-03_15-41-30](https://user-images.githubusercontent.com/596701/38252876-8293a078-3755-11e8-83ba-466d0f105853.png)

jtnord reviewed (commented):
YaY!

scherler commented:
@oleg-nenashev 
![screenshot_2018-04-03_15-43-56](https://user-images.githubusercontent.com/596701/38253038-ecb5d87c-3755-11e8-9dec-e49c81e73473.png) I can add the version footer information again if that is wanted, however it gives away information about the underlying system which under paranoid rules should never be done.

Regarding implementing custom style to the new login page you would only need to "patch" the core login with

     - <link rel="stylesheet" href="${resURL}/css/login.theme.css" type="text/css" />
     + <link rel="stylesheet" href="${resURL}/css/login.myTheme.css" type="text/css" />

scherler commented:
@daniel-beck I am more then willing to update `HudsonIsLoading / HudsonIsRestarting` but would prefer to do that in a separate ticket/PR. @jlong do we have designs for startup/restart?

oleg-nenashev commented:
@scherler 

> Regarding implementing custom style to the new login page you would only need to "patch" the core login with

I think it's important to have a way to replace it in Custom Theme Plugin. Another way would be to allow opting out to a legacy login page by default by using a system property or so.

recena commented on core/src/main/resources/jenkins/model/Jenkins/login.legacy.jelly:
```diff
@@ -0,0 +1,80 @@
+      <st:include page="sidepanel.jelly" />
+    </l:hasPermission>
+    <l:main-panel>
+      <div style="margin: 2em;">
```
Please, no more CSS inline.

scherler replied:
@recena that is the old login.jelly renamed to `login.legacy.jelly ` and changed one line to allow the `loginError.legacy.jelly`

recena commented on core/src/main/resources/jenkins/model/Jenkins/login.legacy.jelly:
```diff
@@ -0,0 +1,80 @@
+      <div style="margin: 2em;">
+        <!-- login form -->
+        <form name="login" action="${it.securityRealm.authenticationGatewayUrl}" method="post" style="text-size:smaller">
+          <table>
```
Please, no more table for implementing layouts. 🐛

scherler replied:
see above that is the old code that I did not changed.

recena replied:
Then, please, do not rename the file.

recena commented on war/src/main/webapp/css/login.css:
```diff
@@ -0,0 +1,108 @@
+ * Create the hover event of the tick
+ */
+.checkbox label:hover::after {
+	opacity: 0.3;
```
Wrong indentation

scherler replied:
will fix

recena commented on war/src/main/webapp/css/login.theme.css:
```diff
@@ -0,0 +1,53 @@
+    color: #4A4A4A;
+}
+.logo {
+    background-image: url('../images/jenkins.svg');
```
We should not use SVG in Jenkins core since it is not fully supported by the HTMLUnit version used in this project. FYI https://github.com/jenkinsci/jenkins-test-harness/pull/95

scherler replied:
Cheers for the headsup. @jlong can you provide a png version?

recena commented on core/src/main/resources/jenkins/model/Jenkins/login.legacy.jelly:
```diff
@@ -0,0 +1,80 @@
+    <st:include it="${it.setupWizard}" page="authenticate-security-token"/>
+  </j:when>
+  <j:otherwise>
+  <l:layout norefresh="true">
```
Why do not we use a [full-screen](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/layout/layout.jelly#L55) layout?

scherler replied:
Again old login screen.

recena replied:
I think I'm not understanding the motivation of this PR 😢 My mistake.

scherler commented:
@oleg-nenashev 

> I think it's important to have a way to replace it in Custom Theme Plugin. Another way would be to allow opting out to a legacy login page by default by using a system property or so.

Well we support request parameter `legacy=true` which will bring you to the `login.legacy.jelly ` which is the same as before. Does that makes sense?

oleg-nenashev commented:
Nope. As a Jenkins admin I would not be able to keep the default behavior

amuniz commented:
> I think it's important to have a way to replace it in Custom Theme Plugin.

The custom theme plugin is evil, however people use it so backward compatibility must be kept.
So I guess leaving the door open (like [it is in layout.jelly](https://github.com/jenkinsci/jenkins/blob/7f311c113bda50862a26534ff50884089da5c41b/core/src/main/resources/lib/layout/layout.jelly#L164-L168)) in the new login.jelly is enough.

scherler committed:
[JENKINS-50447_squash] Fix expression to evaluate when to show an error and when to simply redirect

scherler committed:
eslint - formating changes and fix offences

scherler committed:
[JENKINS-50447_squash] Use system property to controll whether or not to use the legacy login

scherler committed:
[JENKINS-50447_squash] prevent that the jsessionId appears in the url which should fix most tests

scherler commented:
@oleg-nenashev I implemented the legacy fallback now via `SystemPropperties` You can either start with `-Djenkins.model.Jenkins.legacyLogin=true` or set the property with the script console `Jenkins.LEGACY_LOGIN_ENABLED=true`

scherler commented:
@recena first point yes, second point not my main focus.

scherler committed:
[JENKINS-50447_squash] Fix cli test by adding pageDecorators magic from layout.jelly

scherler committed:
[JENKINS-50447_squash] remove pageDecorators because they assume existence of js that we do not have

scherler commented:
The build is green :) I will run the ATH on local system.

scherler committed:
Make theme customisable by using a magic package for css and header. Limiting the implemetation of those to one.

scherler commented:
@oleg-nenashev @amuniz @jtnord @recena please comment on the approach of extensible

batmat commented:
Note: I just built and check locally. We are going to use `/login` to see if it returns an HTTP-200 for part of the _Essentials_ healthchecking process, so it's perfect for now, it still does return 200. Just possibly keep in mind to keep it working.

Apart from that, I tested it, and that looks great! Woot.
![image](https://user-images.githubusercontent.com/223853/38548802-f74ffc92-3cb2-11e8-8548-d5c6bbba41c8.png)

Two NITs springing to mind after some testing: 
* I think clicking on _Keep me signed in_ should check the associated box as it usually does.
* NIT²: I guess using slightly rounded corners would be a bit more consistent with the awesome improvements done recently by Manuel in `/configure`
![image](https://user-images.githubusercontent.com/223853/38548897-2dd3827a-3cb3-11e8-9bf8-00f5e2237905.png)

batmat reviewed (approved):
Not commenting on the code itself, but LGTM from testing: I can still log in, and things look visually fine.

scherler committed:
[JENKINS-50447_squash] Creating a specific decorator for login which only returns one instance

scherler commented:
@batmat hmm not sure why you do not got the rounded corner, looking into that now. Regarding the select button I implemented the design @jlong provided.

I implemented now an alternative way to limit the extensionpoint to one, because of the feedback from @jtnord and @amuniz

scherler committed:
[JENKINS-50447_squash] using rounded corner

recena commented:
I agree with @batmat, we are introducing in Jenkins core the look&feel coming from Blue Ocean. When there is the intention (at least, in my case) of introducing the Bootstrap L&F.

scherler commented:
@batmat and @recena I understand your point. IMO it makes sense that @jlong may review the initial design and align it with the approach you are using.

scherler commented:
Sorry I meant checkbox.

![screenshot_2018-04-10_14-48-42](https://user-images.githubusercontent.com/596701/38557748-49a64cce-3cce-11e8-8dc1-a8b0ce89e0be.png)

There are some "features" in the screenshot that we currently do not have ("I forgot my password...")

scherler commented:
@daniel-beck I am not sure why you added the `WIP` label since I consider the PR `done` unless there is coming new feedback.

scherler committed:
JENKINS-50447 remove legacy code

amuniz commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1771,7 +1770,15 @@ public boolean hyperlinkMatchesCurrentPage(String href) throws UnsupportedEncodi
+     * Gets all the {@link LoginPageDecorator}s.
+     */
+    public static LoginPageDecorator getLoginPageDecorator() {
+        // this method may be called to render start up errors, at which point Hudson doesn't exist yet. see HUDSON-3608
```
Really? AFAICT this is going to be called only by the login screen, so Jenkins is granted to be fully up and running.

I think the `getInstanceOrNull` check can be removed.

amuniz commented on core/src/main/java/jenkins/model/Jenkins.java:
```diff
@@ -472,6 +472,18 @@
+    public boolean isLegacyLoginEnabled() {
+        return Jenkins.LEGACY_LOGIN_ENABLED;
+    }
+
```
I think you forgot to remove this chunk after the review dance, didn't you?

amuniz commented on core/src/main/java/jenkins/model/LoginPageDecorator.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    public static LoginPageDecorator first(){
+        if(Jenkins.getInstanceOrNull()==null) {
```
I think this check is not needed, but if it is then mark the method as `@CheckForNull`

amuniz commented on core/src/main/java/jenkins/model/LoginPageDecorator.java:
```diff
@@ -0,0 +1,32 @@
+import hudson.model.Describable;
+import hudson.model.Descriptor;
+
+public class LoginPageDecorator extends Descriptor<LoginPageDecorator> implements ExtensionPoint, Describable<LoginPageDecorator> {
```
This needs Javadocs

amuniz commented on core/src/main/java/jenkins/model/login/DefaultLoginPageDecorator.java:
```diff
@@ -0,0 +1,8 @@
+import hudson.Extension;
+import jenkins.model.LoginPageDecorator;
+
+@Extension(ordinal=-9999)
```
Add Javadocs please (and please explain why this is `ordinal=-9999` on them)

amuniz commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,119 @@ THE SOFTWARE.
+  <j:set var="extensionsAvailable" value="${h.extensionsAvailable}"/>
+  <j:if test="${request.servletPath=='/' || request.servletPath==''}">
+    ${h.advertiseHeaders(response)}
+    <j:if test="${extensionsAvailable}">
```
I thinks this check is not needed

amuniz commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,119 @@ THE SOFTWARE.
+    <body>
+      <j:choose>
+        <j:when test="${it.setupWizard.isUsingSecurityToken()}">
+          <st:include it="${it.setupWizard}" page="authenticate-security-token"/>
```
Did you check the setup wizard page is not affected by the _new default_ login css?

amuniz commented on core/src/main/resources/jenkins/model/Jenkins/loginError.jelly:
```diff
@@ -24,36 +24,20 @@ THE SOFTWARE.
   <j:choose>
     <j:new var="h" className="hudson.Functions" />
+    <j:when test="${it.isLegacyLoginEnabled()}">
+      <st:include it="${it}" page="loginError.legacy.jelly"/>
```
This does not exist anymore...

amuniz reviewed (changes requested):
There are some leftovers after the review process which need to be removed, and the new extension point needs Javadocs.

batmat reviewed (changes requested):
After a wee bit more thinking, changing my vote to have the click on "Keep me signed in" selecting the checkbox fixed before shipping.

This is the current behaviour when clicking on "Remember me on this computer".

scherler committed:
[JENKINS-50447_squash] Fix comments from @amuniz about javadoc and uniqueness of css for login

scherler committed:
[JENKINS-50447_squash] Fix click on label span to toggle the select of the checkbox

scherler commented:
@batmat I missunderstood what you meant and have fixed that now.

daniel-beck commented:
> The change will provide a new designed login page that is using no external deps to prevent any possible malicious javascript.

> Entry 1: JENKINS-50447, new designed login page that is using no external deps to prevent any possible malicious javascript.

@scherler Could you please explain why this is notable? I understand that web development is an external dependency mess these days, but the existing login page only referenced Jenkins hosted JavaScript, so I don't see why this would be a notable change.

daniel-beck commented:
>  can add the version footer information again if that is wanted, however it gives away information about the underlying system which under paranoid rules should never be done.

This restriction only makes sense when there aren't numerous ways to obtain the information from a locked down Jenkins. There are. So all this does is give a false sense of security-by-obscurity, and makes more difficult to obtain the information legitimately.

daniel-beck commented:
@scherler 

> @daniel-beck I am not sure why you added the `WIP` label since I consider the PR `done` unless there is coming new feedback.

Because that's what was in the title of the PR, so I matched the label to that.

daniel-beck commented:
> @scherler Perhaps we should adapt the `HudsonIsLoading` / `HudsonIsRestarting` UI for consistency?

Considering this some more, I have to give this PR a 👎  in the current state. Not just due to the minor issues listed above, but rather I would *strongly* prefer this change not be done in isolation. I see HudsonIsLoading, HudsonIsRestarting, and the user signup in the same category as the login page, and while this looks great, I can't help but wonder whether it'll saddle us with the inconsistency between them introduced here indefinitely.

daniel-beck commented on core/src/main/java/jenkins/model/LoginPageDecorator.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    /**
+     * The first found LoginDecarator, there can only be one.
```
It is unclear how this is supposed to work in a world with countless plugins. Install a plugin that has `POSITIVE_INFINITY` and everything else loses out?

scherler replied:
The login provides a mean to override some decorators , since we are treating stuff like logo and welcome message they need to be unique. If a plugin provides a custom implementation for the login page then this needs to be unique and yes everyone else loses out, that is the whole point about it.

daniel-beck commented on core/src/main/java/jenkins/model/login/DefaultLoginPageDecorator.java:
```diff
@@ -0,0 +1,14 @@
+package jenkins.model.login;
```
License header.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,120 @@ THE SOFTWARE.
+      <!-- we do not want bots on this page -->
+      <meta name="ROBOTS" content="NOFOLLOW" />
+      <!-- css styling, will fallback to default implementation -->
+      <st:include it="${loginDecorator}" page="login-css.jelly" optional="true"/>
```
Why is the file called `login-css` when it can be any sort of addition to `<head>`?

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,120 @@ THE SOFTWARE.
+              <st:include it="${loginDecorator}" page="login-header.jelly" optional="true"/>
+              <!-- login form -->
+              <form name="login" action="${it.securityRealm.authenticationGatewayUrl}" method="post">
+                <j:if test="${it.securityRealm.allowsSignup()}">
```
Is it deliberate that there is no message at all when signup is disabled?

scherler replied:
https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/jenkins/model/Jenkins/login.jelly#L70-L74

Yeah, if "allowsSignup" is true  then we show the message, in case we do not allow signup, we do not show anything (what should we display?).

daniel-beck replied:
That's okay, just that it referenced login below, and that's still the case otherwise. But if there are no options, it's probably clear what's going on.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,120 @@ THE SOFTWARE.
+                      <input type="checkbox" id="remember_me" name="remember_me" />
+                      <label for="remember_me"> </label>
+                    </div>
+                    <span onClick="checkBoxClick()">${%Keep me signed in}</span>
```
Why is this done with JS and the label left empty?

scherler replied:
we use the label for graphical purposes.

daniel-beck replied:
I don't understand what this means.

Isn't this inaccessible?

scherler replied:
Well the label is only for mouse clicks and not for the ' tab' focus

> This is mentioned in the HTML 4.01 Ref here: http://www.w3.org/TR/html401/interact/forms.html#h-17.9.1
> Specifically near the end of section 17.9.1 and just before 17.10:

    When a LABEL element receives focus, it passes the focus on to its associated control.

daniel-beck replied:
AFAIR, clicking labels (un)checks the associated text box. So why aren't we using the label as it is intended to be used, instead using this JS hack?

scherler replied:
We implemented the same behavior of clicking the label via the onClick method. The styling of the checkbox needed some "dom" tricks to get it working with css. However if that is a big problem for you I will drop the styling of checkbox.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,120 @@ THE SOFTWARE.
+
+              <j:forEach var="fls" items="${it.getFederatedLoginServices()}">
+                <div>
+                  <st:include page="loginFragment.jelly" it="${fls}"/>
```
Will this actually continue to work as expected with the redesign?

daniel-beck replied:
Only existing implementations in https://plugins.jenkins.io/openid and https://plugins.jenkins.io/wwpass-plugin (this one can probably be ignored)

scherler replied:
they should work as before

daniel-beck replied:
Right, but possibly need followup design changes to adapt presentation?

scherler replied:
They may want to adopt their css yeah but https://github.com/jenkinsci/openid-plugin/blob/master/src/main/resources/hudson/plugins/openid/OpenIdLoginService/loginFragment.jelly is providing inline styles, which would be applied.

daniel-beck replied:
So what you're saying is that merging your change will result in broken presentation of the login page downstream in OpenID plugin?

scherler replied:
I just tried but that plugin is broken, anyway all should work as before.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.properties:
```diff
@@ -20,5 +20,7 @@
 # THE SOFTWARE.
 
-signUp=<a href="signup">Create an account</a> if you are not a member yet.
+signUp=Please sign in below or <a href="signup">create an account</a>.
```
Potentially should wipe existing translations.

scherler replied:
Not sure what you mean.

daniel-beck replied:
If there are `login_*.properties` translated files, wipe out the new obsolete translations.

scherler replied:
well, yeah this is how jenkins work, right?

daniel-beck replied:
I am asking you to check `login_*.properties` files for existing translations of this key, and remove them, as they are now probably obsolete and need to be translated again.

scherler replied:
Actually I think not since I only changed the caption of the english translation but the translations of it are not false (only may missing the "sign in below" part but IMO that is not enough to remove the key from the translations.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/loginError.jelly:
```diff
@@ -24,36 +24,17 @@ THE SOFTWARE.
-        </l:main-panel>
-      </l:layout>
+      <j:set var="error" value="true"/>
+      <st:include it="${it}" page="login.jelly"/>
```
I wonder how this impacts SSO type plugins. Do they show the `loginError` page?

scherler replied:
In very rare cases they may end up showing the loginError page however since the auth process is done outside of jenkins, this cases should be limited to a total failure in communication with the sso server, but the angry jenkins page is most likely displayed instead.

daniel-beck reviewed (changes requested):
See previous comments for 👎 reasoning.

Also, we still haven't retired https://wiki.jenkins.io/display/JENKINS/Browser+Compatibility+Matrix so I would expect this change has been tested in those browsers and confirmed working.

scherler commented on core/src/main/java/jenkins/model/LoginPageDecorator.java:
```diff
@@ -0,0 +1,45 @@
+ *
+ * @since TODO
+ */
+public class LoginPageDecorator extends Descriptor<LoginPageDecorator> implements ExtensionPoint, Describable<LoginPageDecorator> {
```
It is like https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/PageDecorator.java#L69

scherler committed:
[JENKINS-50447_squash] fix comments from @daniel-beck

daniel-beck commented on core/src/main/java/jenkins/model/login/DefaultLoginPageDecorator.java:
```diff
@@ -1,3 +1,26 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2004-2009, Sun Microsystems, Inc., Kohsuke Kawaguchi, Erik Ramfelt
```
Really?

scherler committed:
[JENKINS-50447_squash] Fix license headers

scherler committed:
[JENKINS-50447_squash] Implement loading/restarting and signup with same design

scherler commented:
@daniel-beck here some screenshots from the screen you requested to be adopted

### signup

![animated](https://user-images.githubusercontent.com/596701/39205037-de94044e-47f9-11e8-8dd3-a64298537298.gif)

### restart

![animated](https://user-images.githubusercontent.com/596701/39205078-fcc71a8c-47f9-11e8-8072-a1c95dc6f6b4.gif)

### starting

![screenshot_2018-04-24_15-56-26](https://user-images.githubusercontent.com/596701/39205120-1b509618-47fa-11e8-95dd-843a8db44cc5.png)

scherler committed:
[JENKINS-50447_squash] Implement the loading/restarting view and signup with same design as the login page

daniel-beck commented:
Relevant test failures related to signup in HudsonPrivateSecurityRealmTest.

scherler committed:
[JENKINS-50447_squash] Remove limitation that only one validation error can be shown at a time

scherler committed:
[JENKINS-50447_squash] implement password strength checker

daniel-beck commented on core/src/main/resources/hudson/util/HudsonIsLoading/index.jelly:
```diff
@@ -23,22 +23,45 @@ THE SOFTWARE.
+    <j:set var="simpleDecorator" value="${h.simplePageDecorator}"/>
+    <html>
+        <head data-rooturl="${rootURL}" data-resurl="${resURL}" resURL="${resURL}">
+            <title>${h.appendIfNotNull('Create an account!', ' [Jenkins]', 'Jenkins')}</title>
```
Wrong page title

daniel-beck commented on core/src/main/resources/hudson/util/HudsonIsRestarting/index.jelly:
```diff
@@ -23,22 +23,45 @@ THE SOFTWARE.
+    <j:set var="simpleDecorator" value="${h.simplePageDecorator}"/>
+    <html>
+        <head data-rooturl="${rootURL}" data-resurl="${resURL}" resURL="${resURL}">
+            <title>${h.appendIfNotNull('Create an account!', ' [Jenkins]', 'Jenkins')}</title>
```
Wrong page title.

Also, these are fixed strings, `appendIfNotNull` is unnecessary as it will never be null.

daniel-beck reviewed (changes requested):
Looks great overall.

daniel-beck commented:
Would probably also address JENKINS-15355 as there won't be PageDecorators with Adjuncts on HudsonIsLoading anymore.

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,120 @@ THE SOFTWARE.
+  <!-- real deal starting here -->
+  <html>
+    <head data-rooturl="${rootURL}" data-resurl="${resURL}" resURL="${resURL}">
+      <title>${h.appendIfNotNull('simple', ' [Jenkins]', 'Jenkins')}</title>
```
Also terrible page title.

daniel-beck commented on core/src/main/resources/hudson/security/HudsonPrivateSecurityRealm/signup.jelly:
```diff
@@ -26,6 +26,240 @@ THE SOFTWARE.
+                    <div class="signupIntroDefault">
+                        <h1>${%Create an account!}</h1>
+                        <div>If you already have a Jenkins account,
+                            <a href="${rootURL}/login">please sign in</a>
```
No trailing period is inconsistent with the `signUp` message on the sister screen.

daniel-beck commented:
The password strength meter seems a bit like a gimmick as implemented, but I recommend we leave it in and tweak later as needed.

> ![screen shot](https://user-images.githubusercontent.com/1831569/39296415-2153f094-4941-11e8-8318-15f928ec6a7e.png)

daniel-beck commented:
Something appears to be wrong with the restart screen (besides the page title).

> ![screen shot](https://user-images.githubusercontent.com/1831569/39296648-bc325a9c-4941-11e8-960e-3a84fc468f07.png)

scherler commented:
@daniel-beck regarding the strength calculation, I agree and would propose to use e.g. https://lowe.github.io/tryzxcvbn/ as base for score calculation, wdyt?

scherler committed:
[JENKINS-50447_squash] Fix different comments from @daniel-beck regarding titles. Further enhanced spacing of label and input

daniel-beck commented on core/src/main/resources/jenkins/model/Jenkins/login.jelly:
```diff
@@ -23,58 +23,120 @@ THE SOFTWARE.
+  <!-- real deal starting here -->
+  <html>
+    <head data-rooturl="${rootURL}" data-resurl="${resURL}" resURL="${resURL}">
+      <title>${%signIn}</title>
```
:ant: nit -- the consistent `[Jenkins]` suffix everywhere was useful. Just a tab title `sign in` is a bit lightweight…

daniel-beck commented:
As far as I'm concerned, this is done once the issue I observed with the restart screen is addressed 😄  Very excited about this change.

daniel-beck commented:
I tried the latest build and the restart screen is still broken. The screen shot contains some cached stuff, but force-reloading shows completely unstyled text.

Some minor UI findings (that might be deliberate, but noting at least):

- Signup has labels above fields, login has watermarks
- No Jenkins head shown on 'Create account' page?
- "Create an account!" why the exclamation point?
- Login screen has "User" field, signup screen has "Username" for the same thing.
- Is the order of full name first, user name second deliberate? I've seen this in some Linux distros installers (RHEL?) but then they suggest a user name. It's unclear why it works as it does here.

Unaddressed previous review feedback:

- It's unclear why this highlighted the no external JS so strongly (https://github.com/jenkinsci/jenkins/pull/3380#issuecomment-381877994) when we didn't have that before either.
- Still no version on the login page, deliberate or just missed? (https://github.com/jenkinsci/jenkins/pull/3380#issuecomment-381885631)

oleg-nenashev commented:
Will try to test it this weekend

scherler commented:
> I tried the latest build and the restart screen is still broken. The screen shot contains some cached stuff, but force-reloading shows completely unstyled text.

Hmm that should be fixed. What is the page title in that moment. Further can you verify that the page imports 3 css files.

Regarding ui design, the best to answer is @jlong who is the author. I am just the implementor. Imo username should be first and fullname last.

Regarding the version footer I did not see your comment but will address it. Regarding external js libs. Non of the pages justify any use of any lib. Why would I raise loading time of the page for things that are not needed? Let alone I still strongly believe those pages (login and signup) should never have any external deps.

Wadeck reviewed (approved):
Overall it's a really nice re-design, good job. 👍 

Some comments about the UX/UI:
1) The checkboxes in signup and login pages are not consistent

- ![screenshot_2018-04-30_100110_001](https://user-images.githubusercontent.com/2662497/39419339-b2652780-4c5f-11e8-8672-f5b72c71f199.png)

- ![screenshot_2018-04-30_100155_001](https://user-images.githubusercontent.com/2662497/39419340-b28305c0-4c5f-11e8-9419-3f41fc64ebea.png)

2) The labels are not consistently placed between the two pages, inside for the login page and at the top of the inputs for the signup page.

- ![screenshot_2018-04-30_100552_001](https://user-images.githubusercontent.com/2662497/39419354-ce86e188-4c5f-11e8-8275-8b6bd5a7635c.png)

- ![screenshot_2018-04-30_100559_001](https://user-images.githubusercontent.com/2662497/39419355-cea2832a-4c5f-11e8-8dbd-21917b6576f6.png)

3) In signup page the user will fill the "Username"

- ![screenshot_2018-04-30_100626_001](https://user-images.githubusercontent.com/2662497/39419374-dee30e08-4c5f-11e8-9994-6a0adf950604.png)

But then in the login page it's called "User", could be confusing. Perhaps better using "Login" or "Username" in both location.

4) There is a lot larger space between header and form in the signup page than in the login page.

5) The tab order is not user-friendly for the "show password" checkbox that come before the password field (because of its location), but intuitively I expected to be able to "tab" only once after the username to be placed inside the password field.

I like very much the fact that the two pages are very simple and do not provide lots of dependencies.

daniel-beck commented:
@scherler 

> Hmm that should be fixed. What is the page title in that moment. Further can you verify that the page imports 3 css files.

Well, it's not fixed, and easy to reproduce every time. PR build 17: https://www.dropbox.com/s/89lybjp3hrmmbzk/PR3380.mov?dl=0

jlong commented:
@wadatka Good comments. I'll respond inline:

> 1. The checkboxes in signup and login pages are not consistent

Good catch. Yes I agree.

> 2. The labels are not consistently placed between the two pages, inside for the login page and at the top of the inputs for the signup page.

This is intentional. The login page is more minimal. Almost all login forms pair username and password together in the same way, so the explicit labels are unnecessary. However, on a signup form, the fields and ordering are less consistent, so using explicit labels help.

>  3. In signup page the user will fill the "Username"

Yes, it should be "Username" on the login page. Good catch.

>  4. There is a lot larger space between header and form in the signup page than in the login page.

Both forms should be centered vertically on the page. Is this how it appears for you?

> 5. The tab order is not user-friendly for the "show password" checkbox that come before the password field (because of its location), but intuitively I expected to be able to "tab" only once after the username to be placed inside the password field.

That's a good point. I agree. This could be fixed with the `tabindex` property.

/cc @scherler

Wadeck commented:
>  Is this how it appears for you?

See my comment bullet number 2 for the screenshots. Especially the space that is larger on first screenshot.

Thanks for your answers

scherler commented:
https://github.com/jenkinsci/jenkins/pull/3380#issuecomment-385353730 @daniel-beck 
![animated](https://user-images.githubusercontent.com/596701/39754723-7c7b4acc-52c3-11e8-9c9c-dae0021575ff.gif)

daniel-beck commented:
I don't know what to tell you. It still happens for me.

My best guess is that you're on a non-default permissions setup that allows anonymous access, since otherwise the safe redirector doesn't work: The logic has been simplified compared to `hudson-behavior.js` and never leaves the loop when `/` returns 403, as is the case when anonymous does not have Overall/Read.

scherler commented:
@daniel-beck I have not pushed the fix yet, since I am fixing the other comments from the pr. Will push shortly, sorry for the confusion caused.

scherler committed:
[JENKINS-50447_squash] Fix restart screen css. Fix comments regarding signup page.

scherler committed:
[JENKINS-50447_squash] fix tabindex flow

scherler commented:
@Wadeck I fixed your comments, the only thing "the space" is like this on the markups I am using, so for now I have not changed them.

scherler committed:
Merge remote-tracking branch 'upstream/master' into JENKINS-50447_squash

scherler committed:
Merge branch 'master' into JENKINS-50447_squash

scherler committed:
harmonize design
- of checkboxes using jdl styles.
- distance of headers

scherler commented:
@jlong there is an inconsistency in the signup form from the installation wizard and our new page.

Further I strongly feel that the username has to come first and not the fullname since adding an username and submit the form will add this username as fullname.

![screenshot_2018-05-10_13-11-36](https://user-images.githubusercontent.com/596701/39867114-b35974e2-5453-11e8-90fd-02a2a4c69de8.png)

IMO it makes sense to adopt the installation wizard to the new design.

![screenshot_2018-05-10_10-58-54](https://user-images.githubusercontent.com/596701/39867148-d37f21b8-5453-11e8-86c3-ba9f0791acfc.png)

daniel-beck commented:
> IMO it makes sense to adopt the installation wizard to the new design.

When the PR author suggests a giant addition I was too afraid to ask for 😸

jlong commented:
@scherler I do think it would be good to update the form there to be consistent with the sign up form. I would recommend reusing the same fields and updating the "Create first admin user" heading to use the same fonts (it should remain left aligned).

scherler commented:
@daniel-beck and @jlong I am not opposed to do the changes in the setup wizard, however to recap on this PR I started to submit a solution for the login page since then I added 3 more pages. IMHO I compiled with all requirements that had been asked for.me original. Setup is for my a CR aka change request and handled in a follow up, but not here. We missed a couple (literally) of jenkins releases, especially LTS is a bummer to miss, and IMHO there has been no need since this PR is ready to merge before we missed the last release. What is the hold up?

jlong commented:
No hold up on my end. Please merge.

daniel-beck reviewed (changes requested):
Review feedback for the build jenkins-war-2.123-rc15754.b6f4debdc720:

* Having the "Reveal password" checkbox between the password field and the submission field may result in the password being revealed accidentally when tabbing through elements and not expecting this one to be focused, for example:
  > danielbeck(tab)Daniel Beck(tab)s3cr3t(tab)(space to submit)
* The page title `signIn [Jenkins]` should be changed to something that's actually a label to show to users.
* The new 'Restarting Jenkins' page does not handle connection errors while Jenkins is down. Rather than still showing the restart page and attempting again, it will attempt to reload the page and shows an error.

Previous review feedback from https://github.com/jenkinsci/jenkins/pull/3380#issuecomment-384699100 that is still unaddressed (either in code, or responses):

> * No Jenkins head shown on 'Create account' page?
> * "Create an account!" why the exclamation point?
> * Still no version on the login page, deliberate or just missed?

The last one got a response in https://github.com/jenkinsci/jenkins/pull/3380#issuecomment-384717743 that was however not acted on.

daniel-beck commented:
In case there's confusion about the setup wizard change -- I consider this PR complete without it. I think Jenkins would benefit from a standardized look for these screens, but between login, register, and loading, this is a complete package.

I'd still like to see the noted bugs (failing reload, page title) in my 'request changes' comment fixed, and responses or corresponding changes to the other findings.

scherler committed:
Merge remote-tracking branch 'upstream/master' into JENKINS-50447_squash

scherler commented:
< Having the "Reveal password" checkbox between the password field and the submission field may result in the password being revealed accidentally when tabbing through elements and not expecting this one to be focused, for example:

    danielbeck(tab)Daniel Beck(tab)s3cr3t(tab)(space to submit)


The above example ```space to submit)``` is not correct. Space is for changing boolean value. ```ENTER``` is for submit the form (unless you are in a ```textarea```

daniel-beck commented:
It works on Firefox on Mac. Tab to focus, space to trigger. Of course I *could* instead rely on the submit button being defined correctly, but my concern is that this is a valid, potentially common use of form element tabbing/focus, resulting in accidental password exposure.

I don't know whether there's a great solution that leaves the checkbox tab indexable at all:

* Previously, it was before the password field, which was a bit weird.
* Now, it's between password field and submit, which I consider dangerous.
* Moving it to after the submit could impact those who expect the current order and tab twice before pressing space.

Not blocking, but I can't help wonder how many passwords we'll expose this way.

scherler committed:
[JENKINS-50447] Fix comments from @daniel-beck
- fix some accessibility issues
-- html has now a lang tag
-- define main region
- fix missplaced }
- fix when server down for real that we still show the loading page

scherler commented:
> No Jenkins head shown on 'Create account' page?

No, had not been in the design, more information please contact @jlong 

> "Create an account!" why the exclamation point?

Had been in the design, more information please contact @jlong 

> Still no version on the login page, deliberate or just missed?

No, had not been in the design, more information please contact @jlong 

> The page title signIn [Jenkins] should be changed to something that's actually a label to show to users.

Fixed

> The new 'Restarting Jenkins' page does not handle connection errors while Jenkins is down. Rather than still showing the restart page and attempting again, it will attempt to reload the page and shows an error.

Fixed but IMO the prior had been much better UX. I as user would want to know whether Jenkins is dead and unreachable or really restarting. Anyway want to get this PR merged and I am tiered of discussing it.

scherler commented:
regarding the tab order and space keyboard support for form elements, ```SPACE``` is to trigger the underlying element. In case of checkbox is focused it will flip states.

![screenshot_2018-06-06_15-14-56](https://user-images.githubusercontent.com/596701/41040469-cb20a0ce-699c-11e8-8d1e-30b871cf75db.png)

In case of submit button focused it will submit.

![screenshot_2018-06-06_15-15-09](https://user-images.githubusercontent.com/596701/41040507-e5f40102-699c-11e8-8190-01854116cd71.png)

As you can see on the screenshots there are css support to ```hint``` the user what is focused, so I do not see anything to fix there.

scherler committed:
[JENKINS-50447_squash] should never ended up here

daniel-beck commented:
This PR is awaiting responses from @jlong then.

---

> Fixed but IMO the prior had been much better UX. I as user would want to know whether Jenkins is dead and unreachable or really restarting.

The problem is that during an actual restart, Jenkins can easily be unresponsive for long enough that a reload using this code results in a 'cannot be reached' browser error page. That's how I discovered this issue.

jlong commented:
@daniel-beck Having trouble following this thread. What is the issue you need my input on?

daniel-beck commented:
@jlong Thorsten referred me to you for multiple review comments I previously posted in https://github.com/jenkinsci/jenkins/pull/3380#issuecomment-395063308

daniel-beck commented:
I tested the new login screen with OpenID Plugin and it looks reasonable. This is the only implementation of `FederatedLoginService` I could find.

> ![screen shot](https://user-images.githubusercontent.com/1831569/41146708-ad7ed542-6af3-11e8-8dd5-16c2cfbcddd4.png)

daniel-beck commented:
Another review finding (nonblocking):

The password strength indicator moves all form elements. There appears to not be a placeholder of same height before it shows.

jlong commented:
Hi @daniel-beck. Responding to your earlier comments:

> 1) No Jenkins head on 'Create account' page

Yes, intentional. Let's the focus be on the content. More minimal.

It's used on the login page because it is the first point of contact, but is not needed after that.

> 2) Exclamation point on "Create an account!"

Yes, intended to strike a friendly tone.

> 3) Version on the login page

I don't see it as needed, but am open to this if it's really necessary.

daniel-beck reviewed (approved):
No unaddressed review findings from me are left. The last build worked when I tested it, even with OpenID plugin adding content.

I disagree on some details, but those are the things than can be fixed in followup changes based on user feedback.

daniel-beck commented:
> or differently ask what is preventing the merge?

We generally merge PRs towards the end of the week.

daniel-beck commented:
@amuniz's review requesting changes appears to have been addressed.

oleg-nenashev commented:
Please give me some tiime to review it

amuniz commented:
Yup, I'm ok!

agentgonzo commented on core/src/main/java/hudson/Functions.java:
```diff
@@ -1772,7 +1771,14 @@ public boolean hyperlinkMatchesCurrentPage(String href) throws UnsupportedEncodi
-    
+    /**
+     * Gets only one {@link SimplePageDecorator}.
+     * @since FIXME
```
Probably time to fill this in now

agentgonzo commented on core/src/main/java/jenkins/model/SimplePageDecorator.java:
```diff
@@ -0,0 +1,72 @@
+ * <p>
+ * This class provides a few hooks to augment the HTML of the login page.
+ *
+ * @since TODO
```
This could do with filling in too

agentgonzo reviewed (commented):
A couple of TODOs in the `@since` fields could do with filling in

oleg-nenashev reviewed (approved):
It would be great to have a blogpost announcing the change and explaining how to use `SimplePageDecorator`. Changelog entry should also reference Javadoc for this change so that users can follow the link while there is no formal announcement.

The rest looks good to me

mannyci commented:
Hi Guys,

How do I modify the login page with company information etc on the login page with SimplePageDecorator.
Any hints is appreciated.

daniel-beck commented:
@mannyci https://github.com/daniel-beck/login-theme-plugin should help you if you manage to compile it. Will probably end up hosted, just haven't had the time.

(This is not a support forum, and I'll lock this thread in case of further questions. Please use IRC or the mailing lists.)

KostyaSha commented:
Does page miss `shortcut icon` for favicon?

daniel-beck commented:
Caused [JENKINS-69525](https://issues.jenkins.io/browse/JENKINS-69525).
